### ISSUE TITLE
Block smart app install banners

### ISSUE THREAD

**Issue by bbinto** [mozilla-mobile/focus-android#2534]
### Why/User Benefit/User Problem
- Users (who don't want to install apps of the websites they visit) won't get annoyed by having smart app banners show up every single time they go to the website

### What / Requirements
- Setting to block smart app banners

### Acceptance Criteria (how do I know when I’m done?)
- Turn on/off via settings
- Probe to track how many people turn this setting on/off

**Comment by bbinto** [comment]
@brampitoyo to think about UX

**Comment by brampitoyo** [comment]
My consideration for this issue is simply around what to call it.

We can easily call this option something like:

> **Block app install banners**
> Banners that tell you to download and install apps on the Google Play Store

But if we’re using a third-party blocklist – are we, @pocmo? – then I want to make sure that we acknowledge their work in our option.

Maybe something like:

> **Block app install banners**
> Banners that tell you to download and install apps ([block list provided by fanboy.co.nz]())

As for the placement under the “Privacy & Security” settings page, it will be under the “Web Content” subcategory.

What do you think, @BrianNJones?

**Comment by pocmo** [comment]
> But if we’re using a third-party blocklist – are we, @pocmo? 

Yeah. It's unlikely that we will have the resources to maintain our own lists.

**Comment by BrianNJones** [comment]
Let's be as specific as possible. Users may not understand the idea of "app install banners" without all the surrounding context.

[Hedline]
Block websites' app install banners
[Body]
Some websites display a banner prompting you to download their app from the Google Play Store

**Comment by brampitoyo** [comment]
Thanks for the string, @BrianNJones. Good to go.

@pocmo Have we ever initiated contact with the list provider? Should we? I’d like to make sure that they know of what we’re doing, and how their work will be credited in the UI.

**Comment by pocmo** [comment]
> @pocmo Have we ever initiated contact with the list provider? Should we? I’d like to make sure that they know of what we’re doing, and how their work will be credited in the UI.

No contact yet. We should do this as soon as we move this to a specific milestone. In addition to that we require GeckoView APIs (or the WebExtensions hooks) to implement this.

**Comment by bbinto** [comment]
@pocmo I was under the impression we could parse / look up a meta tag in the rendered page. Or does only [iOS support](https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariWebContent/PromotingAppswithAppBanners/PromotingAppswithAppBanners.html) a consistent meta tag?

**Comment by pocmo** [comment]
Yeah, this is a safari-specific feature where it reads the meta tag and shows a "native" app install banner. In our case we'd need to remove/hide all the HTML "Install this app" banners that website show whenever they detect that we are an Android app. :(

**Comment by bbinto** [comment]
@pocmo ok got it, too bad. I thought it was easier. 

Have we done enough research that the fanboy block list is something I could run by Marshall?

Also, assuming you have an approved list, what's the effort including this feature?

**Comment by pocmo** [comment]
I can't estimate this. I'll talk to the GeckoView team to understand how we could support this.

**Comment by samirm** [comment]
what is a smart app supposed to be? PWA?

**Comment by bbinto** [comment]
Hi @samirm - Smart App Banner refers to the prompt that happens on mobile browsers when you get to a website that wants you to download their app.

**Comment by Sdaswani** [comment]
@pocmo how did your chat with the GV team go? https://github.com/mozilla-mobile/focus-android/issues/2534#issuecomment-386533688 Should @boek or @ekager pick it up if you haven't had a chance to talk to them yet?

**Comment by pocmo** [comment]
The result of the discussion was that this is a use case where it makes sense to load a webextension into GeckoView. However support for those APIs is a really big project - definitely not something that will be available in the short term.

**Comment by Sdaswani** [comment]
OK so it sounds like adding this feature is not a great idea until it's ready in GV @bbinto ?

**Comment by bbinto** [comment]
https://news.ycombinator.com/item?id=18416129

**Comment by brampitoyo** [comment]
@bbinto Responding to the HackerNews comment.

The responsible thing for Focus to do seems to be: no page can redirect users to any app, unless they have explicitly interacted with a button or something that will perform the redirect.

So, in the case of – let’s say – a link to the Google Play Store:

1. When you tap on a link, you don’t get redirected to Play Store app. Instead, the Google Play Store website opens in Firefox Focus.
2. If you indeed want to open this web address in the app, then the Google Play Store website will give you a way to do that by tapping “Install”.

In other words, we can rely on pages that likes to open apps automatically, to also carry buttons/banners/links that the user can trigger to open apps manually. So, no worries. The page have solved the app redirection problem for us.
